File: scripts/generate-build-matrix.py
#!/usr/bin/env python3
"""
Generate GitHub Actions build matrix from kernel-matrix.toml.

Usage:
    python scripts/generate-build-matrix.py --matrix kernels/kernel-matrix.toml

Outputs JSON matrix for GitHub Actions strategy.matrix.
"""

import argparse
import json
import sys

try:
    import toml
except ImportError:
    print("Error: toml package required. Install with: pip install toml", file=sys.stderr)
    sys.exit(1)


def generate_matrix(matrix_path: str, runtime_filter: str = "", version_filter: str = "") -> dict:
    """Generate build matrix from TOML configuration."""
    with open(matrix_path) as f:
        config = toml.load(f)

    includes = []

    for runtime_name, runtime_config in config.get("runtimes", {}).items():
        # Apply runtime filter if specified
        if runtime_filter and runtime_name != runtime_filter:
            continue

        versions = runtime_config.get("versions", [])
        architectures = runtime_config.get("architectures", ["x86_64"])

        for version in versions:
            # Apply version filter if specified
            if version_filter and version != version_filter:
                continue

            for arch in architectures:
                includes.append({
                    "runtime": runtime_name,
                    "version": version,
                    "arch": arch,
                })

    return {"include": includes}


def main():
    parser = argparse.ArgumentParser(description="Generate kernel build matrix")
    parser.add_argument("--matrix", required=True, help="Path to kernel-matrix.toml")
    parser.add_argument("--runtime", default="", help="Filter by runtime name")
    parser.add_argument("--version", default="", help="Filter by version")
    args = parser.parse_args()

    matrix = generate_matrix(args.matrix, args.runtime, args.version)

    # Output in GitHub Actions format
    print(f"matrix={json.dumps(matrix)}")


if __name__ == "__main__":
    main()

File: scripts/generate-build-matrix.ts
#!/usr/bin/env bun
/**
 * Generate GitHub Actions build matrix from kernel-matrix.toml.
 *
 * Usage:
 *   bun scripts/generate-build-matrix.ts --matrix kernels/kernel-matrix.toml
 *
 * Outputs JSON matrix for GitHub Actions strategy.matrix.
 */

import { parseArgs } from "util";
import { readFileSync } from "fs";
import { parse as parseToml } from "smol-toml";

interface RuntimeConfig {
  versions: string[];
  architectures?: string[];
  min_memory_mib?: number;
  recommended_memory_mib?: number;
}

interface KernelMatrix {
  unikraft_version?: string;
  defaults?: {
    min_memory_mib?: number;
    recommended_memory_mib?: number;
    boot_args?: string;
  };
  runtimes?: Record<string, RuntimeConfig>;
}

interface MatrixInclude {
  runtime: string;
  version: string;
  arch: string;
}

interface BuildMatrix {
  include: MatrixInclude[];
}

function generateMatrix(
  matrixPath: string,
  runtimeFilter: string = "",
  versionFilter: string = ""
): BuildMatrix {
  const content = readFileSync(matrixPath, "utf-8");
  const config = parseToml(content) as KernelMatrix;

  const includes: MatrixInclude[] = [];

  for (const [runtimeName, runtimeConfig] of Object.entries(
    config.runtimes || {}
  )) {
    // Apply runtime filter if specified
    if (runtimeFilter && runtimeName !== runtimeFilter) {
      continue;
    }

    const versions = runtimeConfig.versions || [];
    const architectures = runtimeConfig.architectures || ["x86_64"];

    for (const version of versions) {
      // Apply version filter if specified
      if (versionFilter && version !== versionFilter) {
        continue;
      }

      for (const arch of architectures) {
        includes.push({
          runtime: runtimeName,
          version,
          arch,
        });
      }
    }
  }

  return { include: includes };
}

function main() {
  const { values } = parseArgs({
    args: process.argv.slice(2),
    options: {
      matrix: {
        type: "string",
      },
      runtime: {
        type: "string",
        default: "",
      },
      version: {
        type: "string",
        default: "",
      },
    },
    strict: true,
    allowPositionals: false,
  });

  if (!values.matrix) {
    console.error("Error: --matrix argument is required");
    process.exit(1);
  }

  const matrix = generateMatrix(
    values.matrix,
    values.runtime || "",
    values.version || ""
  );

  // Output in GitHub Actions format
  console.log(`matrix=${JSON.stringify(matrix)}`);
}

main();
